File: books/physical-ai-humanoid-robotics/code-examples/cpp/foundations/robot_control.cpp
#include <iostream>
#include <memory>
#include "rclcpp/rclcpp.hpp"
#include "geometry_msgs/msg/twist.hpp"
#include "sensor_msgs/msg/laser_scan.hpp"

class RobotController : public rclcpp::Node
{
public:
    RobotController() : Node("robot_controller")
    {
        cmd_vel_publisher_ = this->create_publisher<geometry_msgs::msg::Twist>("/cmd_vel", 10);
        scan_subscription_ = this->create_subscription<sensor_msgs::msg::LaserScan>(
            "/scan", 10,
            std::bind(&RobotController::scan_callback, this, std::placeholders::_1)
        );

        control_timer_ = this->create_wall_timer(
            100ms, std::bind(&RobotController::control_loop, this)
        );

        safe_distance_ = 1.0;
        moving_forward_ = true;
    }

private:
    void scan_callback(const sensor_msgs::msg::LaserScan::SharedPtr msg)
    {
        // Check the front-facing range (middle of the ranges array)
        size_t front_index = msg->ranges.size() / 2;
        float front_scan = msg->ranges[front_index];

        if (front_scan < safe_distance_) {
            moving_forward_ = false;
        } else {
            moving_forward_ = true;
        }
    }

    void control_loop()
    {
        auto cmd_vel_msg = geometry_msgs::msg::Twist();

        if (moving_forward_) {
            cmd_vel_msg.linear.x = 0.5;  // Move forward at 0.5 m/s
            cmd_vel_msg.angular.z = 0.0;
        } else {
            cmd_vel_msg.linear.x = 0.0;  // Stop
            cmd_vel_msg.angular.z = 0.3; // Turn slowly
        }

        cmd_vel_publisher_->publish(cmd_vel_msg);
    }

    rclcpp::Publisher<geometry_msgs::msg::Twist>::SharedPtr cmd_vel_publisher_;
    rclcpp::Subscription<sensor_msgs::msg::LaserScan>::SharedPtr scan_subscription_;
    rclcpp::TimerBase::SharedPtr control_timer_;
    double safe_distance_;
    bool moving_forward_;
};

int main(int argc, char * argv[])
{
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<RobotController>());
    rclcpp::shutdown();
    return 0;
}

File: books/physical-ai-humanoid-robotics/code-examples/python/foundations/robot_control.py
#!/usr/bin/env python3
"""
Basic Robot Control Example
Chapter: Linux + ROS2 Foundations
"""

import rclpy
from rclpy.node import Node
from geometry_msgs.msg import Twist
from sensor_msgs.msg import LaserScan


class RobotController(Node):
    """
    Basic robot controller that moves the robot forward until an obstacle is detected
    """
    def __init__(self):
        super().__init__('robot_controller')
        self.cmd_vel_publisher = self.create_publisher(Twist, '/cmd_vel', 10)
        self.scan_subscriber = self.create_subscription(
            LaserScan,
            '/scan',
            self.scan_callback,
            10
        )

        self.safe_distance = 1.0  # meters
        self.moving_forward = True

        # Timer for control loop
        self.control_timer = self.create_timer(0.1, self.control_loop)

    def scan_callback(self, msg):
        # Check the front-facing range (index around the middle of the array)
        front_scan = msg.ranges[len(msg.ranges) // 2]

        if front_scan < self.safe_distance:
            self.moving_forward = False
        else:
            self.moving_forward = True

    def control_loop(self):
        cmd_vel_msg = Twist()

        if self.moving_forward:
            cmd_vel_msg.linear.x = 0.5  # Move forward at 0.5 m/s
            cmd_vel_msg.angular.z = 0.0
        else:
            cmd_vel_msg.linear.x = 0.0  # Stop
            cmd_vel_msg.angular.z = 0.3  # Turn slowly

        self.cmd_vel_publisher.publish(cmd_vel_msg)


def main(args=None):
    rclpy.init(args=args)
    robot_controller = RobotController()

    try:
        rclpy.spin(robot_controller)
    except KeyboardInterrupt:
        pass
    finally:
        robot_controller.destroy_node()
        rclpy.shutdown()


if __name__ == '__main__':
    main()